Src/ILGPU.Algorithms/ConcurrentStreamProcessor.cs

```diff
@@ -0,0 +1,161 @@
+            Func<Accelerator, AcceleratorStream> streamProvider = null)
+        {
+            maxNumConcurrentStreams = Math.Max(
+                Math.Max(maxNumConcurrentStreams, 1),
+                Environment.ProcessorCount);
+
+            streams = new AcceleratorStream[maxNumConcurrentStreams];
+            if (maxNumConcurrentStreams > 1)
+            {
+                streamProvider ??= accel => accel.CreateStream();
+                for (int i = 0; i < streams.Length; ++i)
+                    streams[i] = streamProvider(accelerator);
+            }
+        }
+
+        #endregion
+
+        #region Properties
+
+        /// <summary>
+        /// Returns the maximum number of concurrent streams supported by this processor.
+        /// </summary>
+        public int MaxNumConcurrentStreams => streams.Length;
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Processes the given action concurrently by using the underlying accelerator
+        /// streams concurrently to submit different jobs in parallel.
+        /// </summary>
+        /// <remarks>
+        /// Note that this method assumes that all previous jobs have been synchronized
+        /// with the current processing thread.
+        /// </remarks>
+        /// <param name="numActions">The number of actions to submit.</param>
+        /// <param name="action">
+        /// The action to invoke on each stream to submit work.
+        /// </param>
+        public void ProcessConcurrently(
+            int numActions,
+            Action<AcceleratorStream, int> action) =>
+            ProcessConcurrently(null, numActions, action);
+
+        /// <summary>
+        /// Processes the given action concurrently by using the underlying accelerator
+        /// streams concurrently to submit different jobs in parallel.
+        /// </summary>
+        /// <param name="stream">The current accelerator stream.</param>
+        /// <param name="numActions">The number of actions to submit.</param>
+        /// <param name="action">
+        /// The action to invoke on each stream to submit work.
+        /// </param>
+        public void ProcessConcurrently(
+            AcceleratorStream stream,
+            int numActions,
+            Action<AcceleratorStream, int> action)
+        {
+            if (numActions < 0)
+                throw new ArgumentOutOfRangeException(nameof(numActions));
+            if (numActions == 0)
+                return;
+            if (action == null)
+                throw new ArgumentOutOfRangeException(nameof(action));
+
+            if (stream != null && numActions == 1 && MaxNumConcurrentStreams == 1)
+            {
+                // Use the given stream to process the action request
+                action(stream, 0);
+            }
+            else
+            {
+                // Wait for the current stream to finish processing
+                stream?.Synchronize();
```

Comment:
This may take a long time waiting for the specified stream, but other streams may be ready and waiting. Basically, my understanding is that locking is bad and you should never do it unless you have to.

Reply:
This is one of the situations in which you have to wait as the async processing streams that will be used afterwards could potentially rely on data written by kernels associated with the input stream.